pkg/resource/deploy/snapshot.go

```diff
@@ -57,6 +58,10 @@ func NewSnapshot(manifest Manifest, secretsManager secrets.Manager,
 //
 // Note: This method modifies the snapshot (and resource.States in the snapshot) in-place.
 func (snap *Snapshot) NormalizeURNReferences() error {
+	// Make a deep copy prior to mutating, since this object's pointer
+	// may be held across thread boundaries as a field on an EngineEvent.
+	var snapCopy = deepcopy.Copy(snap).(*Snapshot)
+	*snap = *snapCopy
```

Comment:
This is a non-atomic write that could appear as a tearing read to another thread, right? We're writing to the same memory location as the original snapshot, but I imagine this appears as a non-atomic write.

Could we modify callers to clone the snapshot into a new memory location before calling `snap.NormalizeURNReferences()` and add a comment that this function mutates and should be provided with an owned copy of the snapshot?

Reply:
I found that this also allowed me to run the test without a race.

It looks like in the two places we call NormalizeURNReferences outside of a test, it's immediately after initializing a new snapshot. If we do the deepcopy inside `NewSnapshot`, that meshes with the name of the method (it's a snapshot!) and I think is logically equivalent to these two lines.

There's one test where we call `NormalizeURNReferences` at the end and verify no error.

```go
// NewSnapshot creates a snapshot from the given arguments.  The resources must be in topologically sorted order.
// This property is not checked; for verification, please refer to the VerifyIntegrity function below.
func NewSnapshot(manifest Manifest, secretsManager secrets.Manager,
	resources []*resource.State, ops []resource.Operation) *Snapshot {

	return &Snapshot{
		Manifest:          deepcopy.Copy(manifest).(Manifest),
		SecretsManager:    deepcopy.Copy(secretsManager).(secrets.Manager),
		Resources:         deepcopy.Copy(resources).([]*resource.State),
		PendingOperations: deepcopy.Copy(ops).([]resource.Operation),
	}
}
```

Reply:
Though I suppose the above is a bit more aesthetically written as:

```go
	snap := &Snapshot{
		Manifest:          manifest,
		SecretsManager:    secretsManager,
		Resources:         resources,
		PendingOperations: ops,
	}
	return deepcopy.Copy(snap).(*Snapshot)
```

Reply:
Nice! I tried that with the snapshot manager, but without success. I thought perhaps the snapshot was leaking from a private field somewhere (i.e. it was assigned from the field into another variable). However, @Frassle 's point still stands. We might not be able to use a deepcopy at all to solve this problem.